### PR TITLE
Us 6

### PR DESCRIPTION
**Pull request by marcint-kainos**
Needs review and possibly testing on one more local enviroment

### CODE REVIEW

File: src/main/resources/assets/js/application/recruiter/directives/tdprAvailabilityDirective.js [hunk 1]
```diff
@@ -9,10 +9,12 @@ define(['angular', 'application/recruiter/tdprRecruiterModule'], function (angul
                     var available = scope.availabilityObject.type ? scope.availabilityObject.type : "unavailable";
                     var type = AvailabilityEnum[available];
 
+                    var elementsCount = scope.$parent.$parent.$parent.timeElementsCount;
```
Comment (jraniszewski) [inline src/main/resources/assets/js/application/recruiter/directives/tdprAvailabilityDirective.js]: Could we checkout length of slots times table ?

File: src/main/resources/assets/js/application/recruiter/services/tdprDateService.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
+define(['angular', 'application/recruiter/tdprRecruiterModule'], function (angular, tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprDateService', function () {
+        this.resetDate = function (dateToReset) {
+            var date = new Date(dateToReset);
+
+            date.setHours(2);
```
Comment (jraniszewski) [inline src/main/resources/assets/js/application/recruiter/services/tdprDateService.js]: Why '2' ?

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+            return updateSlots(
+                reformatSlots(slots, day),
+                personId,
+                formattedData,
+                formattedData
+            );
+        };
+
+        this.prepareAndUpdateSlots = function (slots, personId, day) {
```
Comment (jraniszewski) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: why two different functions instead of one ?

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
+define(['angular', 'application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprDateService'], function (angular, tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprRecruiterSlotsService', ['$http', '$q', 'tdprDateService', 'dateFilter', 'AvailabilityEnum', function ($http, $q, tdprDateService, dateFilter, AvailabilityEnum) {
+        var updateSlots = function(slots, personId, startDate, endDate) {
+            return $http.put("/api/slots/update/" + personId +"/" + startDate + "/" + endDate, slots);
```
Comment (jraniszewski) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: What about handling errors ?

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                formattedData
+            );
+        };
+
+        this.prepareAndUpdateSlots = function (slots, personId, day) {
+            return prepareAndUpdateSlots(slots, personId, day);
+        };
+
+        this.updateSlots = function(slots, personId, startDate, endDate){
+            return updateSlots(slots, personId, startDate, endDate);
```
Comment (jraniszewski) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: why two different functions instead of one ?

File: src/main/resources/assets/stylesheets/recruiter.css [hunk 6]
```diff
@@ -42,7 +42,17 @@
 }
 
 .person-name {
-    min-width: 140px;
+    min-width: 200px;
+    /*position:absolute;
```
Comment (jraniszewski) [inline src/main/resources/assets/stylesheets/recruiter.css]: Is this comment necessary?

File: src/main/resources/assets/js/application/recruiter/directives/tdprAvailabilityDirective.js [hunk 7]
```diff
@@ -1,18 +1,20 @@
...
+                    var available = scope.availabilityObject.type ? scope.availabilityObject.type : "empty";
                     var type = AvailabilityEnum[available];
 
+                    var elementsCount = tdprRecruiterGetSlotsTimesService.getSlotsTimesCount();
+
                     element[0].className = "";
                     element.addClass(type.className);
 
-                    if (scope.availabilityObject.slotId == scope.$parent.$parent.$parent.timeElementsCount) {
+                    if (scope.availabilityObject.index % elementsCount == elementsCount-1) {
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/directives/tdprAvailabilityDirective.js]: use === to compare instead of ==

File: src/main/resources/assets/js/application/recruiter/directives/tdprTableDirective.js [hunk 8]
```diff
@@ -99,7 +106,7 @@ define(['angular', 'application/recruiter/tdprRecruiterModule'], function (angul
                     if (newValue !== undefined) {
                         if (newValue === null) {
                             scope.startWeekDay = previousStartWeekDay;
-                            daysLength = 7;
+                            daysLength = _days.length;
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/directives/tdprTableDirective.js]: Why is this needed? Are we anticipating a change in the number of days?

Comment (marcint-kainos) [inline src/main/resources/assets/js/application/recruiter/directives/tdprTableDirective.js]: This is used for feature for selecting day and then only one day view. Few lines below there is selecting only one day. It can be removed if we won't need it.

Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/directives/tdprTableDirective.js]: I think we decided to not include one-day functionality - the PO said they don't need it

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+                        person: {id: personId},
+                        slot: {id: slots[key].slot},
+                        type: {id: AvailabilityEnum[slotType].id}
+                    });
+                }
+            }
+            return array;
+        };
+
+        this.reformatSlots = function (slots, day, personId) {
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: why are we reformatting slots?

Comment (marcint-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: To work with already existing end-point. Slots we are getting from one endpoint differs from which another end-point expects. It might get changed with refactor, but that's what we ended up doing.

This is what endpoint send us:

```
{
    "day": "2016-08-10",
    "id": 45146,
    "person": 9,
    "slot": 7,
    "type": "available"
}
```

This is what is expected on back-end for slots to update:

```
{
    "person": {
        "id": 11
    },
    "slot": {
        "id": 7
    },
    "slotsDate": "2016-08-08T00:00:00.000Z",
    "type": {
        "id": 2
    }
}
```

Also this function check for day and only sends slots for one day, from whole week.

File: src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
+define(['application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprRecruiterSlotsService', 'application/recruiter/services/tdprDateService'], function (tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprScheduleService', function (tdprRecruiterSlotsService, AvailabilityEnum, tdprDateService) {
+
+        var changeSlotType = function (objectArray, slots, person, changeTo) {
+            var changedFlag = false;
+
+            for (var key in slots) {
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: Why are we constantly using for loops instead of using lodash.js?

File: src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                        break;
+                    }
+                }
+
+                return changeSlotType(objectArray, slots, person, newType);
+            }
+        };
+
+        this.changeSlotType = function (objectArray, slots, person, changeTo) {
+            return changeSlotType(objectArray, slots, person, changeTo)
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: I don't really understand the logic behind this boilerplate code

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterGetSlotsTimesService.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
+define(['angular', 'application/recruiter/tdprRecruiterModule'], function (angular, tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprRecruiterGetSlotsTimesService', ['$http', '$q', function ($http, $q) {
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterGetSlotsTimesService.js]: The name of the service should not contain verbs like Get or Update

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 13]
```diff
@@ -0,0 +1,45 @@
+define(['angular', 'application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprDateService', 'notification'], function (angular, tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprRecruiterSlotsService', ['$http', '$q', 'tdprDateService', 'dateFilter', 'AvailabilityEnum', function ($http, $q, tdprDateService, dateFilter, AvailabilityEnum) {
+        var service = {};
```
Comment (mntss) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: We should probably use this property instead of creating service variable

File: src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
+define(['application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprRecruiterSlotsService', 'application/recruiter/services/tdprDateService'], function (tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprScheduleService', ['tdprRecruiterSlotsService', 'AvailabilityEnum', 'dateFilter', function (tdprRecruiterSlotsService, AvailabilityEnum, dateFilter) {
+        var service = {};
+
+        service.changeSlotType = function (slotId, day, person, changeTo) {
```
Comment (mntss) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: This function should be moved to controller and called from slot directive so we won't need filter functions

File: src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+            person.changesPending = true;
+        };
+
+        service.changeSlotTypeCycleThrough = function (slotId, day, person) {
+            var date = dateFilter(day, "yyyy-MM-dd");
+            var findResult = _.findIndex(person.slotsList, {'day': date, 'number': slotId});
+
+            if (findResult === -1) {
+                // Add available slot for future changes
+                return service.changeSlotType(slotId, date, person, AvailabilityEnum.available.name);
```
Comment (mntss) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: Should recruiter be able to create slot for interviewer with available type?

Comment (marcint-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: This is debatable. Good candidate for issue on our repo and to talk with PO.

File: src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js [hunk 16]
```diff
@@ -0,0 +1,51 @@
+define(['application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprRecruiterSlotsService', 'application/recruiter/services/tdprDateService'], function (tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprScheduleService', ['tdprRecruiterSlotsService', 'AvailabilityEnum', 'dateFilter', function (tdprRecruiterSlotsService, AvailabilityEnum, dateFilter) {
```
Comment (mntss) [inline src/main/resources/assets/js/application/recruiter/services/tdprScheduleService.js]: Should this even be declared as service? There are no data provided

File: src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js [hunk 17]
```diff
@@ -14,22 +16,42 @@ define(['angular', 'application/recruiter/tdprRecruiterModule'
...
+        $scope.showPreviousWeek = function () {
+            offset -= 1;
+            showDataForWeek(offset);
         };
 
-        $scope.showPreviousWeek = function() {
-          offset -= 1;
-          showDataForWeek(offset);
+        $scope.changeSlotTypeCycle = function (slot, slotId, day, personData) {
+            tdprScheduleService.changeSlotTypeCycleThrough(slot, slotId, day, personData);
```
Comment (mntss) [inline src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js]: This function should be in controller not in service

File: src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js [hunk 18]
```diff
@@ -14,22 +15,95 @@ define(['angular', 'application/recruiter/tdprRecruiterModule'
...
-          offset -= 1;
-          showDataForWeek(offset);
+        $scope.changeSlotSubmitChanges = function (personData) {
+            var endDate = new Date($scope.days[4]);
+            endDate.setDate(endDate.getDate() + 1);
+
+            tdprRecruiterSlotsService.updateSlots(personData.slotsList, personData.id, $scope.days[0], endDate).then(
+                function () {
+                    personData.changesPending = false;
+                    Notification.success({message: 'You have successfully submitted data!', delay: 2500});
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js]: This message should be more informative, e.g: Your availability was saved.

File: src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js [hunk 19]
```diff
@@ -14,22 +15,95 @@ define(['angular', 'application/recruiter/tdprRecruiterModule'
...
+                }
+            ).catch(
+                function (status) {
+                    Notification.error({message: status.message, delay: 3500});
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/controllers/tdprWeekTableController.js:104]: Why the different time delays? I think they should all be consistent - or make all success messages shorter, but same, and all error messages longer, but also consistent.

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 20]
```diff
@@ -0,0 +1,30 @@
+define(['angular', 'application/recruiter/tdprRecruiterModule', 'application/recruiter/services/tdprDateService', 'notification'], function (angular, tdprRecruiterModule) {
+    tdprRecruiterModule.service('tdprRecruiterSlotsService', ['$http', '$q', 'tdprDateService', 'dateFilter', 'AvailabilityEnum', function ($http, $q, tdprDateService, dateFilter, AvailabilityEnum) {
+
+        this.updateSlots = function (slots, personId, startDate, endDate) {
+            return $http.put("/api/slots/update/" + personId + "/" + dateFilter(startDate, "dd-MM-yyyy") + "/" + dateFilter(endDate, "dd-MM-yyyy"), this.reformatSlots(slots, personId)).then(function (response) {
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js]: We need to change this endpoint to /api/slots  - without additional verbs - they are defined by the method. And person id should be a query param here.

File: src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+            });
+        };
+
+        this.filterSlots = function (slots, day) {
+            var compareData = dateFilter(day, "yyyy-MM-dd");
+            return _.filter(slots, {"day": compareData});
+        };
+
+        this.reformatSlots = function (slots, personId) {
+            return _.map(slots,
```
Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js:19]: Where is this method being used? Maybe use it in the success callback of the update method so we don't need to use it in controllers.

Comment (marcint-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js:19]: It is used in updateSlots method

```
$http.put("/api/slots/update/" + personId + "/" + dateFilter(startDate, "dd-MM-yyyy") + "/" + dateFilter(endDate, "dd-MM-yyyy"), this.reformatSlots(slots, personId))
```

It is required as long as we don't have other serialising/deserialising methods for updating slots on backend.

Comment (wojciechh-kainos) [inline src/main/resources/assets/js/application/recruiter/services/tdprRecruiterSlotsService.js:19]: 👍
